### ISSUE TITLE
backend: optimize performance

### ISSUE THREAD

**Issue by sanamoin5** [sanamoin5/gadgets#12]
**Objective**: Enhance the efficiency and responsiveness of the system through various optimization techniques.

**Tasks**:

- **Fine-Tune GNN and RL Model Hyperparameters**:
  - **GNN Optimization**:
    - Experiment with different numbers of layers, hidden units, and learning rates.
    - Use validation sets to assess performance improvements.
  - **RL Optimization**:
    - Adjust parameters like learning rate, gamma, and batch size.
    - Experiment with different RL algorithms if necessary.
  
- **Implement Caching Strategies**:
  - **Backend Caching**:
    - Use in-memory caches like [Redis](https://redis.io/) to store frequently accessed data.
    - Example with FastAPI:
      ```python
      from fastapi_cache import FastAPICache
      from fastapi_cache.backends.redis import RedisBackend
      import redis

      redis_client = redis.Redis(host='localhost', port=6379, db=0)
      FastAPICache.init(RedisBackend(redis_client), prefix="fastapi-cache")
      ```
  
  - **Frontend Caching**:
    - Utilize browser caching and service workers to cache static assets.
  
- **Optimize Database Queries**:
  - **Indexing**:
    - Create indexes on frequently queried fields in PostgreSQL and Neo4j.
    - Example for PostgreSQL:
      ```sql
      CREATE INDEX idx_weather_region_id ON weather(region_id);
      ```
  
  - **Query Optimization**:
    - Analyze and optimize slow queries using `EXPLAIN` in PostgreSQL and Neo4j's query profiling tools.
  
- **Improve API Response Times**:
  - **Asynchronous Endpoints**:
    - Utilize FastAPI's asynchronous capabilities to handle I/O-bound operations efficiently.
    - Example:
      ```python
      @app.get("/monitor/status")
      async def read_status():
          return {"status": "API is running"}
      ```
  
  - **Batch Requests**:
    - Combine multiple data fetches into single API calls where possible.
  
- **Optimize Frontend Performance**:
  - **Lazy Loading**:
    - Load map data and other resources only when needed.
  - **Minimize Re-Renders**:
    - Use React's memoization techniques to prevent unnecessary component re-renders.
